docs/specs/app2app.md

```diff
@@ -0,0 +1,199 @@
+# App-to-App authentication
+
+## Abstract
+
+App2app is a mechanism to allow authenticating through another app, which is installed in the same device.
+
+The implementation described in this spec is based on [this blog post](https://openid.net/guest-blog-implementing-app-to-app-authorisation-in-oauth2-openid-connect/).
+
+An app can start the authentication flow by opening a link to another app, instead of using the authorization endpoint. The app which handles the link should validate the authentication request, then could return a valid authorization code which could be used to perform code exchange in oidc.
+
+## Implementation details
+
+### Server-side
+
+- Authorization Endpoint
+
+  - A new parameter `device_key` is supported. This value can be specified by the client to enable app2app login. If specified, it should be a public key of a key pair, which the private key should be stored in a secure storage in the device. Such as the Android keystore or the iOS keychain.
+    - Multiple key types such as RS256 and NIST P-256 are supported.
+
+- Token Endpoint
+
+  - A new grant_type `urn:authgear:params:oauth:grant-type:app2app` is supported. When using such grant_type, The following must be provided in the request:
+    - `refresh_token`: a valid refresh token.
+    - `client_id`: current client id.
+    - `app2app_client_id`: the authenticating client id.
+    - `challenge`: a signature of a challenge token obtained from the `/oauth2/challenge` api, signed with the private key binded with the provided refresh token (using the `device_key` parameter during authentication).
```

Comment:
To follow anonymous user / biometric login, this parameter should be named as `jwt`, and the value is a JWT signed with the private key of the `device_key`. It has the `challenge` claim in the JWT payload.

The JWT header allows us to tell the server which kind of the key is, as well as the signature algorithm. (RS256 or ES256)